File: 0000-0999/0300.LongestIncreasingSubsequence.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ###Binary Search, O(nlgn)
        lis = [0] * len(nums)
        size = 0
        for num in nums:
            i, j = 0, size
            while i != j:
                m = (i + j) / 2
                if lis[m] < num:
                    i = m + 1
                else:
                    j = m
            lis[i] = num
            if i == size:
                size += 1
        return size

        '''# AC, but what about [1, 10, 20, 2, 3, 4] ?
        lis = 0
        for i in xrange(len(nums)):
            temp = 1
            pre_pre = pre = nums[i]
            for j in xrange(i+1, len(nums)):
                if nums[j] > pre:
                    temp += 1
                    pre_pre = pre
                    pre = nums[j]
                elif pre_pre < nums[j] < pre:
                    pre = nums[j]
            lis = max(lis, temp)
        return lis
        '''

File: 0000-0999/0300.LongestIncreasingSubsequence.go
func lengthOfLIS(nums []int) int {
    // binSearch, O(nlgn) + O(n)
    inc := make([]int, 0, len(nums))
    for _, num := range nums {
        idx := binInsertLeft(inc, num)
        if idx == len(inc) {
            inc = append(inc, num)
        } else {
            inc[idx] = num
        }
    }
    return len(inc)
    /* DP, O(n^2) + O(n)
    n := len(nums)
    dp := make([]int, n)
    ans := 0
    for i := 0; i < n; i++ {
        dp[i] = 1
        for j := 0; j < i; j++ {
            if nums[i] > nums[j] {
                dp[i] = max(dp[i], dp[j]+1)
            }
        }
        ans = max(ans, dp[i])
    }
    return ans
    */
}

func binInsertLeft(nums[]int, target int) int {
    l, r := 0, len(nums)
    for l < r {
        mid := (l+r) / 2
        if nums[mid] < target {
            l = mid + 1
        } else {
            r = mid
        }
    }
    return l
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}